### ISSUE TITLE
64核CPU；RTX4090；三并发卡顿问题请教

### ISSUE THREAD

**Issue by WangSiyao666** [lipku/LiveTalking#516]
服务器配置为：
CPU：64核；
显卡：NVIDIA GeForce RTX 4090；

![Image](https://github.com/user-attachments/assets/73783a74-d3f2-4073-a8e9-6f3c1ccb1d88)

使用以下命令启动（musetalkV1.5 + fish-speech）
--transport webrtc --model musetalk --avatar_id avator_3 --max_session 3 --tts fishtts --TTS_SERVER http://127.0.0.1:8080 --REF_FILE test 
只开一个正常CPU占用800

<img width="1936" height="272" alt="Image" src="https://github.com/user-attachments/assets/2217ff08-2fed-4bea-9a02-8bdb4f2ffd07" />
开第二个，两个一起朗读

<img width="1926" height="350" alt="Image" src="https://github.com/user-attachments/assets/61669fd9-02c4-497a-a558-ad6e750d0cd5" />
开第三个，三个一起朗读

<img width="1931" height="326" alt="Image" src="https://github.com/user-attachments/assets/e540780a-24bd-4700-bf04-3415c04f4286" />

最后看了一下GPU，24G显存基本占满了。

<img width="1008" height="256" alt="Image" src="https://github.com/user-attachments/assets/a2f6a1e6-96e9-4af7-bd9c-3fc573713279" />

请问是什么原因导致的卡顿，是显存不够么？如果是显存的问题，那么可以配置使用多卡么？

**Comment by heyyyyou** [comment]
可能因为每个启动后都会有一个渲染模型在显存中占用，然后三个加起来你的显存就不够了。尝试自己改下共用一个渲染模型试试

**Comment by WangSiyao666** [comment]
> 可能因为每个启动后都会有一个渲染模型在显存占用，然后三个加起来你的显存就不够了。尝试自己改下模拟一个渲染模型试用
好的老师我去尝试一下，然后我又发现了一个情况，现在我所有的连接都已经断开了，并且页面也关闭了，但是显存并没有释放。
<img width="1015" height="447" alt="Image" src="https://github.com/user-attachments/assets/960da1ef-2987-4d3f-a553-0846cc752806" />

**Comment by WangSiyao666** [comment]
> 遇到同样的问题，Tesla T4（16G显存）、112核心CPU、跑了6路，再多一路就开始明显显卡顿，且开始语音延迟，或者语音播放不全的现象，另外当出现这个问题之后，减少并行路数，发现显存也不会恢复。

@EminemJK 请问老师，您是怎么做到16G显存跑6路的？是 --max_session 设置了6 么？我现在24G的显存，只启动项目就占了14个G，在加上TTS 以及 ASR 显存就所剩无几了

**Comment by heyyyyou** [comment]
> 在我所有的连接都已经断开了，并且页面也关闭了，但是显存并没有释放。

可能缓存，后面一直占用不释放吗

**Comment by WangSiyao666** [comment]
> > 在我所有的连接都已经断开了，并且页面也关闭了，但是显存并没有释放。
> 
> 可能缓存，后面一直占用不释放吗

@heyyyyou 是的老师，到现在显存一直没有释放。

<img width="1006" height="256" alt="Image" src="https://github.com/user-attachments/assets/ae7478ad-5835-4434-a222-d5b1160c0d60" />

**Comment by heyyyyou** [comment]
> > > 在我所有的连接都已经断开了，并且页面也关闭了，但是显存并没有释放。
> > 
> > 
> > 可能缓存，后面一直占用不释放吗
> 
> [@heyyyyou](https://github.com/heyyyyou) 是的老师，到现在显存一直没有释放。
> 
> <img alt="Image" width="1006" height="256" src="https://private-user-images.githubusercontent.com/167976085/477351993-ae7478ad-5835-4434-a222-d5b1160c0d60.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUwNzIyODcsIm5iZiI6MTc1NTA3MTk4NywicGF0aCI6Ii8xNjc5NzYwODUvNDc3MzUxOTkzLWFlNzQ3OGFkLTU4MzUtNDQzNC1hMjIyLWQ1YjExNjBjMGQ2MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxM1QwNzU5NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMzViMzg3ZDU4M2M3MTA4N2NlZDM2MGM1Yzk0ODgwZTNiMzdlZTJiZTNlMmU3NTYzMWNkYWEwNmIxYmIzYWU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.p92l9rY89HBV1gHwpf0XMfrdOPrDj9ZdHxu_nz_Tiao">

感觉更像是 启动后的musetalk模型占用，项目启动后会 加载MuseTalk模型组件（VAE、UNet、PE）加载avatar数据 执行模型预热（warm_up）这些模型组件在启动时就被加载到GPU显存中，并且会持续占用显存直到程序完全退出。所以就算你关了连接，但是项目仍然在执行中，这种设计是为了性能优化而做出的权衡 - 保持模型在显存中可以避免重复加载的开销，但代价是显存持续占用。

（不用叫老师 大家互相学习）

**Comment by 5dprince** [comment]
用的docker吗?

**Comment by WangSiyao666** [comment]
> 用的docker吗?

@5dprince 没用docker

**Comment by EminemJK** [comment]
> > 遇到同样的问题，Tesla T4（16G显存）、112核心CPU、跑了6路，再多一路就开始明显显卡顿，且开始语音延迟，或者语音播放不全的现象，另外当出现这个问题之后，减少并行路数，发现显存也不会恢复。
> 
> [@EminemJK](https://github.com/EminemJK) 请问老师，您是怎么做到16G显存跑6路的？是 --max_session 设置了6 么？我现在24G的显存，只启动项目就占了14个G，在加上TTS 以及 ASR 显存就所剩无几了

我这边情况大概是这样，docker 部署，训练素材视频720P，max_session=100，其他代码均源码未修改。另外我的跑6路是指同时打开6个tab播放出人物，要是同时推理的话，应该不到6路。这个并发是个难题。

**Comment by WangSiyao666** [comment]
> > 卡顿很有可能是你显卡每秒能推理的图片数量超过过了你所有并发数量的fps之和，如果你显卡每秒推理100张图片，每一路视频流25fps，那么理论上可以应对4路并发。这时候你再开一路视频，就会导致推理的帧率跟不上，从而出现卡顿。
> 
> 显存在模型不变的前提下，只和每次推理的batch_size大小相关。

我在musereal.py中做了一些打印
```Python
@torch.no_grad()
def inference(render_event,batch_size,input_latent_list_cycle,audio_feat_queue,audio_out_queue,res_frame_queue,
              vae, unet, pe,timesteps): #vae, unet, pe,timesteps
    length = len(input_latent_list_cycle)
    index = 0
    count=0
    counttime=0
    logger.info('start inference')
    while render_event.is_set():
        try:
            whisper_chunks = audio_feat_queue.get(block=True, timeout=1)
        except queue.Empty:
            continue
        is_all_silence=True
        audio_frames = []
        for _ in range(batch_size*2):
            frame,type,eventpoint = audio_out_queue.get()
            audio_frames.append((frame,type,eventpoint))
            if type==0:
                is_all_silence=False
        if is_all_silence:
            for i in range(batch_size):
                res_frame_queue.put((None,__mirror_index(length,index),audio_frames[i*2:i*2+2]))
                index = index + 1
        else:
            t=time.perf_counter()
            t0 = time.perf_counter()
            whisper_batch = np.stack(whisper_chunks)
            latent_batch = []
            for i in range(batch_size):
                idx = __mirror_index(length,index+i)
                latent = input_latent_list_cycle[idx]
                latent_batch.append(latent)
            latent_batch = torch.cat(latent_batch, dim=0)
            audio_feature_batch = torch.from_numpy(whisper_batch)
            audio_feature_batch = audio_feature_batch.to(device=unet.device,
                                                            dtype=unet.model.dtype)
            audio_feature_batch = pe(audio_feature_batch)
            latent_batch = latent_batch.to(dtype=unet.model.dtype)
            torch.cuda.synchronize()
            t1 = time.perf_counter()

            pred_latents = unet.model(latent_batch,
                                        timesteps, 
                                        encoder_hidden_states=audio_feature_batch).sample
            torch.cuda.synchronize()
            t2 = time.perf_counter()

            recon = vae.decode_latents(pred_latents)

            torch.cuda.synchronize()
            t3 = time.perf_counter()

            logger.info(f"data_to_gpu={t1 - t0:.4f}s, unet={t2 - t1:.4f}s, vae={t3 - t2:.4f}s")

            counttime += (time.perf_counter() - t)
            count += batch_size
            if count>=100:
                logger.info(f"------actual avg infer fps:{count/counttime:.4f}")
                count=0
                counttime=0
            for i,res_frame in enumerate(recon):
                res_frame_queue.put((res_frame,__mirror_index(length,index),audio_frames[i*2:i*2+2]))
                index = index + 1
    logger.info('musereal inference processor stop')
```
如果是用Thread模式进行启动的话，使用4090显卡，三路并发的时候：unet=0.9 vae=0.4 基本是这样的
显卡显存占用了大概有19G，还能剩余5G。但是显卡的负载会间歇性跳满，应该就是在处理图片的时候就会跳满。所以我可不可以理解成MuseTalk V1.5 这个模型在4090显卡上实时推理的极限基本也就是这样了。

**Comment by ahaofan** [comment]
> > > 卡顿很有可能是你显卡每秒能推理的图片数量超过过了你所有并发数量的fps之和，如果你显卡每秒推理100张图片，每一路视频流25fps，那么理论上可以应对4路并发。这时候你再开一路视频，就会导致推理的帧率跟不上，从而出现卡顿。
> > 
> > 
> > 显存在模型不变的前提下，只和每次推理的batch_size大小相关。
> 
> 我在musereal.py中做了一些打印
> 
> @torch.no_grad()
> def inference(render_event,batch_size,input_latent_list_cycle,audio_feat_queue,audio_out_queue,res_frame_queue,
>               vae, unet, pe,timesteps): #vae, unet, pe,timesteps
>     length = len(input_latent_list_cycle)
>     index = 0
>     count=0
>     counttime=0
>     logger.info('start inference')
>     while render_event.is_set():
>         try:
>             whisper_chunks = audio_feat_queue.get(block=True, timeout=1)
>         except queue.Empty:
>             continue
>         is_all_silence=True
>         audio_frames = []
>         for _ in range(batch_size*2):
>             frame,type,eventpoint = audio_out_queue.get()
>             audio_frames.append((frame,type,eventpoint))
>             if type==0:
>                 is_all_silence=False
>         if is_all_silence:
>             for i in range(batch_size):
>                 res_frame_queue.put((None,__mirror_index(length,index),audio_frames[i*2:i*2+2]))
>                 index = index + 1
>         else:
>             t=time.perf_counter()
>             t0 = time.perf_counter()
>             whisper_batch = np.stack(whisper_chunks)
>             latent_batch = []
>             for i in range(batch_size):
>                 idx = __mirror_index(length,index+i)
>                 latent = input_latent_list_cycle[idx]
>                 latent_batch.append(latent)
>             latent_batch = torch.cat(latent_batch, dim=0)
>             audio_feature_batch = torch.from_numpy(whisper_batch)
>             audio_feature_batch = audio_feature_batch.to(device=unet.device,
>                                                             dtype=unet.model.dtype)
>             audio_feature_batch = pe(audio_feature_batch)
>             latent_batch = latent_batch.to(dtype=unet.model.dtype)
>             torch.cuda.synchronize()
>             t1 = time.perf_counter()
> 
>             pred_latents = unet.model(latent_batch,
>                                         timesteps, 
>                                         encoder_hidden_states=audio_feature_batch).sample
>             torch.cuda.synchronize()
>             t2 = time.perf_counter()
> 
>             recon = vae.decode_latents(pred_latents)
> 
>             torch.cuda.synchronize()
>             t3 = time.perf_counter()
> 
>             logger.info(f"data_to_gpu={t1 - t0:.4f}s, unet={t2 - t1:.4f}s, vae={t3 - t2:.4f}s")
> 
>             counttime += (time.perf_counter() - t)
>             count += batch_size
>             if count>=100:
>                 logger.info(f"------actual avg infer fps:{count/counttime:.4f}")
>                 count=0
>                 counttime=0
>             for i,res_frame in enumerate(recon):
>                 res_frame_queue.put((res_frame,__mirror_index(length,index),audio_frames[i*2:i*2+2]))
>                 index = index + 1
>     logger.info('musereal inference processor stop')
> 如果是用Thread模式进行启动的话，使用4090显卡，三路并发的时候：unet=0.9 vae=0.4 基本是这样的 显卡显存占用了大概有19G，还能剩余5G。但是显卡的负载会间歇性跳满，应该就是在处理图片的时候就会跳满。所以我可不可以理解成MuseTalk V1.5 这个模型在4090显卡上实时推理的极限基本也就是这样了。

只要你显卡利用率爆满,基本上算力就到上限了，和剩余的显存关系不大，瓶颈在算力上。我4090D 48G显存的卡，每秒最多推理100张图片，已经到了上限。
